Dockerfile

```diff
@@ -3,69 +3,99 @@
 # default LISTEN port to 3000
 ARG PORT=3000
 
-FROM node:16-slim as dev
+#
+# 1. install only deps (to be cached)
+#
+FROM node:16-bullseye-slim as deps
 LABEL Unlock <ops@unlock-protocol.com>
 
 # args need to be mentioned at each stage
 ARG BUILD_DIR
 ARG PORT
 
-# setup home dir
-RUN mkdir -p /home/unlock
-RUN chown -R node /home/unlock
-WORKDIR /home/unlock
-
-# Setting user as root to handle apk install
+# Setting user as root to handle apt install
 USER root
 
-# apk steps merged to leverage virtual install of package
-# allowing for removal after yarn dependencies install
+# install all deps required to build modules
 RUN apt-get update \
  && DEBIAN_FRONTEND=noninteractive \
     apt-get install --no-install-recommends --assume-yes \
     bash \
     git \
     python3 \
-    build-essential \
+    build-essential
+
+# switch to user
+ENV DEST_FOLDER=/opt/manifests
+
+RUN mkdir $DEST_FOLDER
+
+# copy files 
+COPY . .
+
+# copy all package.json files
+RUN find . -maxdepth 3 -mindepth 2 -type f -name "package.json" ! -path '**node_modules**' ! -path './docker' -exec cp --parents '{}' $DEST_FOLDER \;
+COPY package.json $DEST_FOLDER
+COPY yarn.lock $DEST_FOLDER
+
+# copy al yarn related files
+COPY .yarnrc.yml $DEST_FOLDER
+COPY .prettierrc $DEST_FOLDER
+COPY .yarn/plugins $DEST_FOLDER/.yarn/plugins
+COPY .yarn/releases $DEST_FOLDER/.yarn/releases
+RUN chown -R node:node $DEST_FOLDER
+
+# specify yarn cache folder location (to be used by docker buildkit)
+RUN echo "cacheFolder: ${DEST_FOLDER}/yarn-cache" >> $DEST_FOLDER/.yarnrc.yml
+RUN mkdir $DEST_FOLDER/yarn-cache
+RUN chown -R node:node $DEST_FOLDER/yarn-cache
+
+# install deps
+WORKDIR ${DEST_FOLDER}
+RUN yarn
+
+#
+# 2. build packages and prepare image for testing/dev
+#
+FROM node:16-bullseye-slim as dev
+
+# install all deps required to build packages
+RUN apt-get update \
```

Comment:
I think these change very little, so maybe move upward?

Reply:
we just copy the file from the stage above, so we have to install things to compile deps

Reply:
actually better to build everything on first step (see https://github.com/unlock-protocol/unlock/pull/10241#issuecomment-1311623124)